server/src-lib/Hasura/RQL/DML/Internal.hs

```diff
@@ -273,18 +273,19 @@ dmlTxErrorHandler = mkTxErrorHandler $ \case
     , PGInvalidColumnReference ]
   _ -> False
 
-toJSONableExp :: Bool -> PGColumnType -> S.SQLExp -> S.SQLExp
-toJSONableExp strfyNum colTy expn
+toJSONableExp :: Bool -> PGColumnType -> Bool -> S.SQLExp -> S.SQLExp
+toJSONableExp strfyNum colTy asText expn
   | isScalarColumnWhere isGeoType colTy =
       S.SEFnApp "ST_AsGeoJSON"
       [ expn
       , S.SEUnsafe "15" -- max decimal digits
       , S.SEUnsafe "4"  -- to print out crs
       ] Nothing
       `S.SETyAnn` S.jsonTypeAnn
-  | isScalarColumnWhere isBigNum colTy && strfyNum =
-      expn `S.SETyAnn` S.textTypeAnn
-  | otherwise = expn
+  | isScalarColumnWhere isBigNum colTy && strfyNum = asTextExpn
+  | otherwise = if asText then asTextExpn else expn
+  where
+    asTextExpn = expn `S.SETyAnn` S.textTypeAnn
```

Comment:
I think it would probably be slightly clearer to write this:
```haskell
| asText || (isScalarColumnWhere isBigNum colTy && strfyNum) =
    expn `S.SETyAnn` S.textTypeAnn
| otherwise = expn
```